nexus/src/nexus.rs

```diff
@@ -2249,6 +2259,49 @@ impl Nexus {
     pub async fn session_hard_delete(&self, token: String) -> DeleteResult {
         self.db_datastore.session_hard_delete(token).await
     }
+
+    pub async fn updates_refresh_metadata(&self) -> Result<(), Error> {
+        let rack = self.as_rack();
+        let trust_root = self
+            .tuf_trusted_root
+            .as_ref()
+            .ok_or_else(|| Error::InvalidRequest {
+                message: "updates system not configured".into(),
+            })?
+            .clone();
+
+        let artifacts = tokio::task::spawn_blocking(move || {
+            crate::updates::read_artifacts(&rack, &trust_root)
+        })
+        .await
+        // first, the JoinError:
+        .map_err(|e| Error::InternalError { internal_message: e.to_string() })?
+        // next, the boxed dyn Error:
+        .map_err(|e| Error::InternalError {
+            internal_message: e.to_string(),
+        })?;
+
+        // FIXME: if we hit an error in any of these database calls, the available artifact table
+        // will be out of sync with the current artifacts.json. can we do a transaction or
+        // something?
```

Comment:
> can we do a transaction or something

Yeah, there's an example here: https://github.com/oxidecomputer/async-bb8-diesel/blob/22c26ef840075a57b18ac2eae3ead989b992773e/examples/usage.rs#L72-L82

If we wanted do this, we would probably have a call to `self.db_datastore` that takes all the artifacts we wanna upsert at once, so the transaction could be self-contained in the datastore function.

All that being said... the upserts are idempotent, right? We *could* have a separate single-column table for `targets_role_version`, which we update once all the necessary artifacts are updated. Then the steps would be:

1. upsert all the artifacts
2. update the "desired" targets role version
3. delete all older artifacts

And I don't *think* they'd need to be transactional anymore.

Either way, if you wanna punt on this until after the PR is merged, could we file a bug to track? DB concurrency shenanigans are hairy, but I don't want us to lose track of 'em.

Reply:
How is this function called?  If it's from an API endpoint, and the API request fails, then at least the caller knows it didn't work and we're basically saying its their responsibility to retry until it works.

Is it also a problem that it might be inconsistent?  That is, we might have updated some of the artifacts but not all of them?

Reply:
If it's inconsistent, then either Nexus will be missing information about a newer update, or might try to update to a removed artifact (if we ever have to do that for some reason).